pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -24,307 +24,480 @@
-    if (_storageType == StorageType.ARRAY_BASED) {
-      return new ArrayBasedGroupKeyIterator();
-    } else {
-      final ObjectIterator<Map.Entry<Long, Integer>> iterator = _groupKeyToId.entrySet().iterator();
-      return new MapBasedGroupKeyIterator(iterator);
+    switch (_storageType) {
+      case ARRAY_BASED:
+        return new ArrayBasedGroupKeyIterator();
+      case LONG_MAP_BASED:
+        return new LongMapBasedGroupKeyIterator();
+      case ARRAY_MAP_BASED:
+        return new ArrayMapBasedGroupKeyIterator();
+      default:
+        throw new RuntimeException("Unsupported storage type.");
     }
   }
 
   /**
    * {@inheritDoc}
-   * @return
    */
   @Override
   public int getNumGroupKeys() {
-    return _numUniqueGroupKeys;
+    return _numGroupKeys;
   }
 
   /**
-   * Save the passed in group key:
-   * - For ARRAY_BASED storage, we store unique keys in a boolean array
-   *   (key is indexed into the array).
-   * - For MAP_BASED storage, we generate a unique new key, which is the total
-   *   number of unique keys, and use that as the group key. We also store the
-   *   mapping between the original group by key and the newly generated index.
+   * Helper function to get the group key associated to the long type raw key. If there is no one, generate a new group
+   * key for this raw key and put them into the map.
+   * (LONG_MAP_BASED storage type)
    *
-   * @param rawKey
-   * @return
+   * @param rawKey raw key.
+   * @return group key.
    */
   private int updateRawKeyToGroupKeyMapping(long rawKey) {
-    int groupKey;
-    if (_storageType == StorageType.ARRAY_BASED) {
-      int intRawKey = (int) rawKey;
-      if (!_uniqueGroupKeysFlag[intRawKey]) {
-        _uniqueGroupKeysFlag[intRawKey] = true;
-        _numUniqueGroupKeys++;
+    int groupKey = _groupKeyToId.get(rawKey);
+    if (groupKey == INVALID_ID) {
+      groupKey = _numGroupKeys++;
+      _groupKeyToId.put(rawKey, groupKey);
+    }
+    return groupKey;
+  }
+
+  /**
+   * Helper function to get the group key associated to the IntArrayList type raw key. If there is no one, generate a
+   * new group key for this raw key and put them into the map.
+   * (ARRAY_MAP_BASED storage type)
+   *
+   * @param rawKey raw key.
+   * @return group key.
+   */
+  private int updateRawKeyToGroupKeyMapping(IntArrayList rawKey) {
+    int groupKey = _arrayGroupKeyToId.getInt(rawKey);
+    if (groupKey == INVALID_ID) {
+      groupKey = _numGroupKeys++;
+      _arrayGroupKeyToId.put(rawKey, groupKey);
+    }
+    return groupKey;
+  }
+
+  /**
+   * Helper function to generate group keys (int[]) according to the document id. This method should only be called when
+   * there are multi value group-by columns.
+   * (ARRAY_BASED storage type)
+   *
+   * @param index index of the docIdSet.
+   * @param docId document id.
+   * @return group keys.
+   */
+  private int[] generateKeysForDocIdArrayBased(int index, int docId) {
```

Comment:
The three flavors of this method seem to be doing very similar things? Any way to factor out portions of reusable code?